Objects/longobject.c

```diff
@@ -4511,16 +4510,17 @@ long_rshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
         if (newsize <= 0)
             return PyLong_FromLong(0);
         hishift = PyLong_SHIFT - remshift;
-        lomask = ((digit)1 << hishift) - 1;
-        himask = PyLong_MASK ^ lomask;
         z = _PyLong_New(newsize);
         if (z == NULL)
             return NULL;
-        for (i = 0, j = wordshift; i < newsize; i++, j++) {
-            z->ob_digit[i] = (a->ob_digit[j] >> remshift) & lomask;
-            if (i+1 < newsize)
-                z->ob_digit[i] |= (a->ob_digit[j+1] << hishift) & himask;
```

Comment:
The new code also fixes a _really_ subtle portability issue in this code. Here the result of `a->ob_digit[j+1] * 2**hishift` may not be representable in the target type. Normally that wouldn't matter, because `a->ob_digit[j+1]` has type `digit`, which is an unsigned type, so the C standards tell us that any out-of-range value wraps in the normal way. But integer promotions could result in the left-hand operand _actually_ being of type `int` (a signed type), and then an out-of-range shift result gives undefined behaviour according to the standard. We don't run into this in practice because under any likely combination of integer type bit widths (e.g., 16-bit `digit`, 32-bit `int`), if `digit` is small enough to be promoted to `int`, then `int` is likely big enough to hold the shift result. But the C standard does allow potentially problematic bit widths (e.g., `digit` could be 16 bits and `int` 24 bits).

Not a real issue, since it's unlikely we'd ever meet this in practice, but it's nice not to have to worry about it. With the new code, the result of the shift is guaranteed to be representable in the target type (that type being either `twodigits`, or something larger in the case that there are integer promotions going on).